lib/Backend/arm64/EncoderMD.cpp

```diff
@@ -939,6 +965,38 @@ EncoderMD::GenerateEncoding(IR::Instr* instr, BYTE *pc)
         bytes = this->EmitMovConstant(Emitter, instr, EmitMovz, EmitMovz64);
         break;
 
+    case Js::OpCode::MOVK_SHIFT:
+    {
+        Assert(instr->GetSrc1()->IsLabelOpnd());
+        Assert(instr->GetSrc2()->IsIntConstOpnd());
+        IR::LabelInstr* labelInstr = instr->GetSrc1()->AsLabelOpnd()->GetLabel();
+        uint32 shift = instr->GetSrc2()->AsIntConstOpnd()->AsUint32();
+
+        // We're going to drop src2, set src1 to just the masked bits from the label
+        // offset (so we don't even need to go into relocation), and emit it.
+        instr->UnlinkSrc2();
+        uintptr_t fullvalue = labelInstr->GetOffset();
+        instr->ReplaceSrc1(IR::IntConstOpnd::New((fullvalue & (0xffff << shift)) >> shift, IRType::TyUint16, instr->m_func, true));
+        bytes = this->EmitMovConstantKnownShift(Emitter, instr, EmitMovk, EmitMovk64, shift);
+    }
+        break;
+
+    case Js::OpCode::MOVZ_SHIFT:
+    {
+        Assert(instr->GetSrc1()->IsLabelOpnd());
+        Assert(instr->GetSrc2()->IsIntConstOpnd());
+        IR::LabelInstr* labelInstr = instr->GetSrc1()->AsLabelOpnd()->GetLabel();
+        uint32 shift = instr->GetSrc2()->AsIntConstOpnd()->AsUint32();
+
+        // We're going to drop src2, set src1 to just the masked bits from the label
+        // offset (so we don't even need to go into relocation), and emit it.
+        instr->UnlinkSrc2();
+        uintptr_t fullvalue = labelInstr->GetOffset();
+        instr->ReplaceSrc1(IR::IntConstOpnd::New((fullvalue & (0xffff << shift)) >> shift, IRType::TyUint16, instr->m_func, true));
+        bytes = this->EmitMovConstantKnownShift(Emitter, instr, EmitMovz, EmitMovz64, shift);
+    }
+        break;
+
     case Js::OpCode::MRS_FPCR:
```

Comment:
I debated whether to make MOVK/MOVN/MOVZ take the shift explicitly. Looks like you found a good reason. I would rather there be only one form of these and you change all the other instances to encode the shift, rather than adding more opcodes.

Reply:
Ok, I've set the short-forms to take shifts, and removed the _SHIFT versions.